internal/datastore/mysql/revisions.go

```diff
@@ -0,0 +1,178 @@
+	// implementation deviates slightly from PSQL implementation in order to support
+	// database seeding in runtime, instead of through migrate command
+	ctx, span := tracer.Start(ctx, "HeadRevision")
+	defer span.End()
+
+	revision, err := mds.loadRevision(ctx)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+	if revision == 0 {
+		return datastore.NoRevision, nil
+	}
+
+	return revisionFromTransaction(revision), nil
+}
+
+func (mds *Datastore) CheckRevision(ctx context.Context, revision datastore.Revision) error {
+	// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+	ctx, span := tracer.Start(ctx, "CheckRevision")
+	defer span.End()
+
+	revisionTx := transactionFromRevision(revision)
+
+	freshEnough, future, err := mds.checkValidTransaction(ctx, revisionTx)
+	if err != nil {
+		return fmt.Errorf(errCheckRevision, err)
+	}
+
+	if !freshEnough {
+		return datastore.NewInvalidRevisionErr(revision, datastore.RevisionStale)
+	}
+	if future {
+		return datastore.NewInvalidRevisionErr(revision, datastore.RevisionInFuture)
+	}
+
+	return nil
+}
+
+func (mds *Datastore) loadRevision(ctx context.Context) (uint64, error) {
+	// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+	// slightly changed to support no revisions at all, needed for runtime seeding of first transaction
+	ctx, span := tracer.Start(ctx, "loadRevision")
+	defer span.End()
+
+	query, args, err := mds.GetLastRevision.ToSql()
+	if err != nil {
+		return 0, fmt.Errorf(errRevision, err)
+	}
+
+	var revision sql.NullInt64
+	err = mds.db.QueryRowContext(datastore.SeparateContextWithTracing(ctx), query, args...).Scan(&revision)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return 0, nil
+		}
+		return 0, fmt.Errorf(errRevision, err)
+	}
+	if !revision.Valid {
+		// there are no rows in the relation tuple transaction table
+		return 0, nil
+	}
+
+	return uint64(revision.Int64), nil
+}
+
+func (mds *Datastore) checkValidTransaction(ctx context.Context, revisionTx uint64) (bool, bool, error) {
+	ctx, span := tracer.Start(ctx, "checkValidTransaction")
+	defer span.End()
+
+	var freshEnough, future sql.NullBool
+
+	err := mds.db.QueryRowContext(
+		datastore.SeparateContextWithTracing(ctx), mds.validTransactionQuery, revisionTx, revisionTx,
+	).Scan(&freshEnough, &future)
+	if err != nil {
+		return false, false, fmt.Errorf(errCheckRevision, err)
+	}
+
+	span.AddEvent("DB returned validTransaction checks")
+
+	return freshEnough.Bool, future.Bool, nil
+}
+
+func (mds *Datastore) createNewTransaction(ctx context.Context, tx *sql.Tx) (newTxnID uint64, err error) {
+	ctx, span := tracer.Start(ctx, "createNewTransaction")
+	defer span.End()
+
+	createQuery := mds.createTxn
+	if err != nil {
+		return 0, fmt.Errorf("createNewTransaction: %w", err)
+	}
+
+	result, err := tx.ExecContext(ctx, createQuery)
+	if err != nil {
+		return 0, fmt.Errorf("createNewTransaction: %w", err)
+	}
+
+	lastInsertID, err := result.LastInsertId()
```

Comment:
I did some digging into the MySQL driver implementation of `LastInsertId()` ([source](https://github.com/go-sql-driver/mysql/blob/ad9fa14acdcf7d0533e7fbe58728f3d216213ade/connection.go#L316-L322)):

```go
// mysqlConn struct
type mysqlConn struct {
	// ... some stuff
	affectedRows     uint64
	insertId         uint64
	// more stuff ...
}
```

```go
// Exec implementation
// ... some stuff
err := mc.exec(query)
if err == nil {
	return &mysqlResult{
		affectedRows: int64(mc.affectedRows),
		insertId:     int64(mc.insertId),
	}, err
}
// more stuff ...
```

Internally, the MySQL driver uses a `uint64` to store `insertedId`. It will convert that to an `int64` to return from `LastInsertId()`. This lead me to wonder if there would be overflow issues here. So I tested in the Go playground: https://go.dev/play/p/m6qQJtcotAe. As the playground shows, since the original value was a `uint64`, we can convert the `int64` to a `uint64` without a problem.

Reply:
Awesome, thanks for digging in 👍